File: sistema_esperto/KB_personality.pl
:- dynamic ha_tratto/3.  % ha_tratto(Persona, Tratto, Valore)

tipo_personalita(introverso).
tipo_personalita(estroverso).

% ----- PESI -----

% tempo_da_solo
peso(introverso, tempo_da_solo, alto, 0.9).
peso(introverso, tempo_da_solo, medio, 0.5).
peso(introverso, tempo_da_solo, basso, 0.1).

peso(estroverso, tempo_da_solo, alto, 0.1).
peso(estroverso, tempo_da_solo, medio, 0.4).
peso(estroverso, tempo_da_solo, basso, 0.9).

% paura_palco
peso(introverso, paura_palco, si, 0.9).
peso(introverso, paura_palco, no, 0.2).

peso(estroverso, paura_palco, si, 0.1).
peso(estroverso, paura_palco, no, 0.8).

% partecipazione_eventi
peso(introverso, partecipazione_eventi, alta, 0.2).
peso(introverso, partecipazione_eventi, media, 0.5).
peso(introverso, partecipazione_eventi, bassa, 0.9).

peso(estroverso, partecipazione_eventi, alta, 0.9).
peso(estroverso, partecipazione_eventi, media, 0.6).
peso(estroverso, partecipazione_eventi, bassa, 0.2).

% frequenza_uscite
peso(introverso, frequenza_uscite, alta, 0.2).
peso(introverso, frequenza_uscite, media, 0.5).
peso(introverso, frequenza_uscite, bassa, 0.9).

peso(estroverso, frequenza_uscite, alta, 0.9).
peso(estroverso, frequenza_uscite, media, 0.6).
peso(estroverso, frequenza_uscite, bassa, 0.1).

% svuotato_dopo_socializzazione
peso(introverso, svuotato_dopo_socializzazione, si, 0.9).
peso(introverso, svuotato_dopo_socializzazione, no, 0.3).

peso(estroverso, svuotato_dopo_socializzazione, si, 0.2).
peso(estroverso, svuotato_dopo_socializzazione, no, 0.8).

% dimensione_amici
peso(introverso, dimensione_amici, grande, 0.2).
peso(introverso, dimensione_amici, medio, 0.5).
peso(introverso, dimensione_amici, piccolo, 0.9).

peso(estroverso, dimensione_amici, grande, 0.9).
peso(estroverso, dimensione_amici, medio, 0.6).
peso(estroverso, dimensione_amici, piccolo, 0.2).

% frequenza_post
peso(introverso, frequenza_post, alta, 0.3).
peso(introverso, frequenza_post, media, 0.6).
peso(introverso, frequenza_post, bassa, 0.9).

peso(estroverso, frequenza_post, alta, 0.9).
peso(estroverso, frequenza_post, media, 0.6).
peso(estroverso, frequenza_post, bassa, 0.2).

% ----- COMPORTAMENTI -----
evita_socializzazione(Persona) :- 
    ha_tratto(Persona, partecipazione_eventi, bassa);
    ha_tratto(Persona, svuotato_dopo_socializzazione, si).

ama_stare_solo(Persona) :- 
    ha_tratto(Persona, tempo_da_solo, alto);
    ha_tratto(Persona, frequenza_uscite, bassa);
    ha_tratto(Persona, dimensione_amici, piccolo).

a_suo_agio(Persona) :- 
    ha_tratto(Persona, partecipazione_eventi, alta);
    ha_tratto(Persona, svuotato_dopo_socializzazione, no).

ama_compagnia(Persona) :- 
    ha_tratto(Persona, tempo_da_solo, basso);
    ha_tratto(Persona, frequenza_uscite, alta);
    ha_tratto(Persona, dimensione_amici, grande).

comunicazione_digitale_bassa(Persona) :- 
    ha_tratto(Persona, frequenza_post, bassa);
    ha_tratto(Persona, partecipazione_eventi, bassa).

comunicazione_digitale_alta(Persona) :- 
    ha_tratto(Persona, frequenza_post, alta);
    ha_tratto(Persona, partecipazione_eventi, alta).

riservatezza(Persona) :- 
    ama_stare_solo(Persona);
    comunicazione_digitale_bassa(Persona).

tendenza_sociale_alta(Persona) :- 
    comunicazione_digitale_alta(Persona);
    ama_compagnia(Persona).

% ----- CLASSIFICAZIONE TRATTI -----
comportamenti_introversi([evita_socializzazione, ama_stare_solo, comunicazione_digitale_bassa, riservatezza]).
comportamenti_estroversi([a_suo_agio, ama_compagnia, comunicazione_digitale_alta, tendenza_sociale_alta]).

% ----- CONTEGGIO -----
conteggio_veri(_, [], 0).
conteggio_veri(Persona, [Comportamento|Resto], Conteggio) :-
    (call(Comportamento, Persona) -> 
        conteggio_veri(Persona, Resto, ConteggioResto), 
        Conteggio is ConteggioResto + 1
    ; 
        conteggio_veri(Persona, Resto, Conteggio)
    ).

conteggio_comportamenti(Persona, introverso, N) :-
    comportamenti_introversi(ListaComportamenti),
    conteggio_veri(Persona, ListaComportamenti, N).

conteggio_comportamenti(Persona, estroverso, N) :-
    comportamenti_estroversi(ListaComportamenti),
    conteggio_veri(Persona, ListaComportamenti, N).

% ----- PUNTEGGIO -----
% somma pesi per ogni tipo
punteggio(Persona, Tipo, Totale) :-
    findall(Peso, (ha_tratto(Persona, Tratto, Valore), peso(Tipo, Tratto, Valore, Peso)), ListaPesi), media(ListaPesi, Totale).

% media dei pesi
media(Lista, Media) :- sum_list(Lista, Somma), length(Lista, Len), Len > 0, Media is Somma / Len.

% ----- CLASSIFICAZIONE -----
classificazione(Persona, Tipo, CountIntro, CountEstro, P1, P2) :-     
    % conteggio dei tratti
    conteggio_comportamenti(Persona, introverso, CountIntro),
    conteggio_comportamenti(Persona, estroverso, CountEstro),
    punteggio(Persona, introverso, P1),
    punteggio(Persona, estroverso, P2),

    Diff is abs(CountIntro - CountEstro),
    (Diff > 1 -> 
        (CountIntro > CountEstro -> Tipo = introverso; Tipo = estroverso)
    ;

    % logica a punteggio alternativa (situazione bilanciata)
    (P1 > P2 -> Tipo = 'tendente a introverso'; Tipo = 'tendente a estroverso')
    ).

File: sistema_esperto/KB_personality.py
from pyswip import Prolog

# funzione principale per input dell'utente, interrogazione della KB e output
def run_personality_expert_system():
    prolog = Prolog()
    try:
        prolog.consult("KB_personality.pl")
    except Exception as e:
        print(f"ERRORE: Impossibile caricare 'KB_personality.pl'. Assicurati che il file sia nella stessa cartella.")
        print(f"Dettaglio errore: {e}")
        return
    
    # funzione per risposta si/no
    def ask_yes_no(question):
        while True:
            risposta = input(f"{question} (si/no): ").lower()
            if risposta in ['si', 'no']:
                return risposta
            print("ERRORE: Inserimento non valido, per favore inserisci 'si' o 'no'.")

    # funzione per risposta multipla
    def ask_multiple_choice(question, options):
        while True:
            print(f"\n{question} Opzioni possibili: {', '.join(options)}")
            risposta = input("La tua scelta: ").lower()
            if risposta in options:
                return risposta
            print("ERRORE: Inserimento non valido, per favore scegli una delle opzioni elencate.")

    # funzione per risposta numerica
    def ask_numeric(question):
        while True:
            risposta_str = input(f"{question} ").strip()
            try:
                numero = int(risposta_str)
                return numero
            except ValueError:
                print("ERRORE: Inserimento non valido, per favore inserisci un numero intero.")

    # inizio dialogo con sistema esperto
    print("-----Sistema Esperto per la Personalità-----")
    print("\nQuesto sistema sarà in grado di determinare la tua personalità (introverso oppure estroverso).\nRispondi alle domande per ottenere una previsione.\n")

    # domande con risposta si/no
    paura_palco = ask_yes_no("Provi paura da palcoscenico?")
    prolog.assertz(f"ha_tratto(utente, paura_palco({paura_palco}))")

    svuotato_dopo_socializzazione = ask_yes_no("\nTi senti svuotato dopo aver socializzato?")
    prolog.assertz(f"ha_tratto(utente, svuotato_dopo_socializzazione({svuotato_dopo_socializzazione}))")

    # domande con risposta multipla
    partecipazione_eventi = ask_multiple_choice(
        "Con quale frequenza partecipi a eventi sociali?",
        ["alta", "media", "bassa"])
    prolog.assertz(f"ha_tratto(utente, partecipazione_eventi({partecipazione_eventi}))")

    frequenza_uscite = ask_multiple_choice(
        "Con quale frequenza esci?",
        ["alta", "media", "bassa"])
    prolog.assertz(f"ha_tratto(utente, frequenza_uscite({frequenza_uscite}))")

    frequenza_post = ask_multiple_choice(
        "Con quale frequenza pubblichi post sui social media?",
        ["alta", "media", "bassa"])
    prolog.assertz(f"ha_tratto(utente, frequenza_post({frequenza_post}))")

    # domande con risposta numerica
    tempo_da_solo = ask_numeric("\nQuante ore al giorno trascorri da solo in media?")
    categoria_tempo = "alto" if tempo_da_solo >= 8 else "medio" if tempo_da_solo >= 4 else "basso"
    prolog.assertz(f"ha_tratto(utente, tempo_da_solo({categoria_tempo}))")
        
    dimensione_amici = ask_numeric("\nQuante persone ci sono nella tua cerchia di amici più stretta?")
    categoria_amici = "grande" if dimensione_amici >= 8 else "medio" if dimensione_amici >= 4 else "piccolo"
    prolog.assertz(f"ha_tratto(utente, dimensione_amici({categoria_amici}))")
    
    # interrogazione della KB per determinare la personalità
    print("\n--- Analisi della personalità in corso... ---")
    try:
        results = list(prolog.query("personalita_con_punteggio(utente, Tipo, Punteggio)"))

        if not results:
            print("Nessuna personalità ha ottenuto un punteggio. Risultato: INDETERMINATO")
            return
                        
        best_result = max(results, key=lambda x: x['Punteggio'])
        max_score = best_result['Punteggio']
        winners = [res['Tipo'] for res in results if res['Punteggio'] == max_score]

        # sistema a punteggio: se il punteggio è uguale, si ha la parità
        if len(winners) == 1:
            final_prediction = str(winners[0]).upper()
        else:
            winner_names = [str(w).upper() for w in winners]
            final_prediction = f"INDETERMINATO (Parità tra: {', '.join(winner_names)})"

        print(f"\n--- Risultato finale: ---")
        print(f"\nLa personalità dell'individuo è: {final_prediction}")
    
    except Exception as e:
        print(f"ERRORE durante l'interrogazione della KB: {e}")

if __name__ == "__main__":
    run_personality_expert_system()
        